src/main/java/org/junitpioneer/jupiter/Range.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2015-2018 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import org.junit.platform.commons.util.PreconditionViolationException;
+import org.junit.platform.commons.util.Preconditions;
+
+/**
+ * An iterator for numerical ranges, used as the backing logic for {@link RangeSourceProvider}.
+ * @param <N> The numerical type used by the range.
+ */
+public abstract class Range<N extends Number & Comparable<N>> implements Iterator<N> {
+	private N from;
+	private N to;
+	private N step;
+	private boolean closed;
+	private N current;
+	private int sign;
+
+	Range(N from, N to, N step, boolean closed) {
+		this.from = from;
+		this.to = to;
+		this.step = step;
+		this.closed = closed;
+		current = null;
+		sign = step.compareTo(getZero());
+	}
+
+	/**
+	 * Asserts the range is valid.
+	 * @throws PreconditionViolationException if the range is not valid
+	 */
+	void validate() throws PreconditionViolationException {
+		Preconditions.condition(!step.equals(getZero()), "Illegal range. The step cannot be zero.");
+
+		Preconditions.condition(closed || !from.equals(to),
+			"Illegal range. Equal from and to will produce an empty range.");
+
+		int cmp = from.compareTo(to);
+		Preconditions.condition((cmp < 0 != sign < 0) || (closed && cmp == 0),
+			() -> String.format("Illegal range. There's no way to get from %s to %s with a step of %s.", from, to,
+				step));
+	}
+
+	N getStep() {
+		return step;
+	}
+
+	N getCurrent() {
+		return current;
+	}
+
+	/**
+	 * The next value in the range. Calling {@link #next()} will return this value and advance the iterator to it.
+	 */
+	abstract N nextValue();
+
+	private N getNextValue() {
+		if (current == null) {
+			return from;
+		}
+		return nextValue();
+	}
+
+	/**
+	 * The value of the no-op "zero", illegal step in terms of N
+	 */
+	abstract N getZero();
+
+	@Override
+	public boolean hasNext() {
+		N nextValue = getNextValue();
+		int cmp = nextValue.compareTo(to);
+		return (cmp < 0 != sign < 0) || (closed && cmp == 0);
+	}
+
+	@Override
+	public N next() {
+		if (!hasNext()) {
+			throw new NoSuchElementException();
+		}
+		current = getNextValue();
+		return current;
+	}
+
+	Stream<N> asStream() {
+		return StreamSupport.stream(Spliterators.spliteratorUnknownSize(this, Spliterator.ORDERED), false);
+	}
```

Comment:
I'd prefer it this way, smaller API surface:

```java
	Stream<N> asStream() {
		return StreamSupport.stream(Spliterators.spliteratorUnknownSize(new ValueIterator(), Spliterator.ORDERED), false);
	}

	private class ValueIterator implements Iterator<N> {

		@Override
		public boolean hasNext() {
			N nextValue = getNextValue();
			int cmp = nextValue.compareTo(to);
			return (cmp < 0 != sign < 0) || (closed && cmp == 0);
		}

		@Override
		public N next() {
			if (!hasNext()) {
				throw new NoSuchElementException();
			}
			current = getNextValue();
			return current;
		}

	}
```

Reply:
Just to be clear - if you're suggesting remove `asStream` from `Range`, sure, that's not a problem, will do.

If you're suggesting separating `Range` from a new `ValueIterator` class, I'm not sure I'm following.

Reply:
Yes, I was suggesting to make `ValueIterator` an inner class, so that `Range` does not implement `Iterator`. One week later, I don't really see a problem with that, though. An iterable range makes sense. Let's leave it as it is. :)